### PR TITLE
ENH: Adds filter_requires_grad.

### LINKED ISSUE #260: Add utility function to filter non-grad parameters from optimizer
See #236.

For freezing certain layers it would be helpful to mark the freezed parameters by setting `requires_grad=False` and have a utility function filter those weights before passing all weights to the optimizer.

For example:

```python
def filtered_optimizer(pgroups, **kwargs):
    params = skorch.utils.filter_requires_grad(pgroups)
    return torch.optim.SGD(params, **kwargs)

net = Net(module, optimizer=filtered_optimizer)
```

### PR DESCRIPTION
**Pull request by thomasjpfan**
Resolves #260.

`filter_requires_grad` accepts a list of pgroups dictionaries and yields a list of pgroups with the parameters that have `requires_grad==False` filtered out. All other values in the dictionaries are left alone.

### CODE REVIEW

File: skorch/tests/test_utils.py [hunk 1]
```diff
@@ -417,3 +417,72 @@ def type_truth_table():
...
+        assert len(list((filter_pgroups[1]['params']))) == 1
+
+        assert filter_pgroups[0]['lr'] == 0.1
+
+    def test_some_params_requires_graident(
+            self, filter_requires_grad):
+        pgroups = [{
+            'params': [
+                torch.zeros(1, requires_grad=True),
+                torch.zeros(1, requires_grad=True)
```
Comment (ottonemo) [inline skorch/tests/test_utils.py]: I think we should test wether the filter function does not simply drop whole groups where only one element is filtered. We could do this by setting requires_grad to False here as well.

File: skorch/utils.py [hunk 2]
```diff
@@ -415,3 +415,22 @@ def store_step(self, step):
...
         """Return the stored step."""
         return self.step
+
+
+def filter_requires_grad(pgroups):
+    """Returns parameter groups where parameters with
+    ``requires_grad==False`` are filtered out.
+    """
+    for pgroup in pgroups:
+        output = {k: v for k, v in pgroup.items() if k != 'params'}
```
Comment (ottonemo) [inline skorch/utils.py]: Why not use `params = pgroup.pop('params')`?

Comment (thomasjpfan) [reply skorch/utils.py]: As a utility function, I choose not to mutate elements in `pgroups`.

File: skorch/tests/test_utils.py (hunk 1 above)
Comment (thomasjpfan) [reply skorch/tests/test_utils.py]: I agree. PR was updated with this test.

File: skorch/tests/test_utils.py [hunk 3]
```diff
@@ -417,3 +417,90 @@ def type_truth_table():
...
+
+
+class TestFilterRequiresGrad():
+
+    @pytest.fixture
+    def filter_requires_grad(self):
+        from skorch.utils import filter_requires_grad
+        return filter_requires_grad
+
+    def test_all_parameters_requires_graident(
```
Comment (benjamin-work) [inline skorch/tests/test_utils.py]: Typo: `graident` -> `gradient`

File: skorch/tests/test_utils.py (hunk 1 above)
Comment (benjamin-work) [reply skorch/tests/test_utils.py]: Where is that tested?

**Comment by thomasjpfan** [comment]
I'm going to put this on hold till the details on how to use this is laid out in   https://github.com/dnouri/skorch/issues/261.

I would prefer these functions be hidden from the user, and `NeutralNet` be able to filter out `require_grads==False` be default.

**Comment by benjamin-work** [comment]
> I would prefer these functions be hidden from the user, and NeutralNet be able to filter out require_grads==False be default.

In general, I would agree that this the more sensible default. pytorch, however, does explicitely not do it, and according to the principle of least surprise, I lean towards going with pytorch.

Still, it should be easy for the user to change the behavior and a helper function + docs is probably sufficient.

**Comment by benjamin-work** [comment]
Sorry that I haven't thought of it earlier, but could you please add a test that shows that if I initialize a net with additional optimizer arguments, the arguments are correctly passed to the optimizer:

`NeuralNet(..., optimizer=filtered_optimizer(torch.optim.SGD), optimizer__momentum=0.9)`

File: skorch/tests/test_utils.py (hunk 1 above)
Comment (thomasjpfan) [reply skorch/tests/test_utils.py]: Ah I misunderstood the comment, I updated this `test_some_params_requires_gradient` to reflect this.

**Comment by thomasjpfan** [comment]
Update this PR with:

1. Adds test to check that arguments are correctly passed to the optimizer. (`test_passes_kwargs_to_neuralnet_optimizer`)
2. Moves functions to `skorch.helper`.

File: skorch/helper.py [hunk 4]
```diff
@@ -78,3 +79,13 @@ def __repr__(self):
     @property
     def shape(self):
         return (self._len,)
+
+
+def filtered_optimizer(optimizer):
```
Comment (taketwo) [inline skorch/helper.py]: As discussed in #261, this should probably have a different name. Besides, I think docstring can be more descriptive. Right now without checking the code it's not obvious whether `optimizer` argument should be a class or an instantiated object.

**Comment by thomasjpfan** [comment]
I updated this PR with your suggestions.

If `optimizer__filter_requires_grad` gets accepted as a parameter to `NeutralNet`, these functions would be moved to util, and a user would not need to worry about importing `filter_requires_grad`.

**Comment by benjamin-work** [comment]
All in all, this looks good, but I'm not yet satisfied with the name `filter_requires_grad`. First it sounds as if it were a "filter that requires gradients". Second, it's not really a filter but a wrapper. A more telling name would be `optimizer_with_only_params_that_require_grad`, which obviously is too long. So more along `optimizer_params_requires_grad`?

**Comment by ottonemo** [comment]
I think we can avoid the naming issue by generalising, e.g.:
```
def filtered_optimizer(optimizer, filter_fn):
    der opt(params, **kwargs):
        return optimizer(filter_fn(params), **kwargs)
    return opt
```

**Comment by benjamin-work** [comment]
Thanks, good work.
